prov/net/src/xnet_progress.c

```diff
@@ -128,21 +229,28 @@ static ssize_t xnet_send_msg(struct xnet_ep *ep)
 
 	assert(xnet_progress_locked(xnet_ep2_progress(ep)));
 	assert(ep->cur_tx.entry);
+	if (!xnet_get_io_uring_credit(&xnet_ep2_progress(ep)->tx_io_uring))
+		return -FI_EAGAIN;
+
 	tx_entry = ep->cur_tx.entry;
 	ret = ofi_bsock_sendv(&ep->bsock, tx_entry->iov, tx_entry->iov_cnt,
 			      &len);
-	if (ret < 0 && ret != -FI_EINPROGRESS)
+	if (ret >= 0)
+		len = ret;
+	else if (ret == -FI_EIOURING_PREP) {
+		assert(!ep->cur_tx.io_uring_busy);
+		ep->cur_tx.io_uring_busy = true;
 		return ret;
-
-	if (ret == -FI_EINPROGRESS) {
+	}
+	else if (ret == -FI_EINPROGRESS) {
```

Comment:
Okay, so you need to distinguish between 2 async 'error' codes based on whether using io_uring or async sockets.